File: 数仓表结构/monitor/src/main/java/com/weshare/bigdata/util/SendEmailUtil.java
package com.weshare.bigdata.util;

import org.apache.commons.lang3.StringUtils;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.mail.*;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
import java.io.IOException;
import java.io.InputStream;
import java.util.Date;
import java.util.Properties;

/**
 * @author yuheng.wang
 * @date 2020/10/21 15:47
 * @Description
 */
public class SendEmailUtil {
    private static final Logger LOGGER = LoggerFactory.getLogger(SendEmailUtil.class);
    private static InputStream inputStream = null;
    private static Properties prop = null;

    public static Session initSession() throws NoSuchProviderException, IOException {
        inputStream = ClassLoader.getSystemResourceAsStream("email.properties");
        prop = new Properties();
        prop.load(inputStream);
        inputStream.close();
        return Session.getInstance(prop);
    }

    public static void sendEmail(String subjectInfo,String content){
        Transport transport = null;
        try {
            //open Transport
            Session session = initSession();
            String senderAccount = prop.getProperty("senderAccount");
            String senderPassword = prop.getProperty("senderPassword");
            LOGGER.warn("senderAccount name is : " + senderAccount);
            transport = session.getTransport();
            //create transport connection
            if (!StringUtils.isBlank(senderAccount) && !StringUtils.isBlank(senderPassword)) {
                transport.connect(senderAccount,senderPassword);
                LOGGER.warn("transport connect success!");
            } else {
                transport.connect();
                LOGGER.warn("transport connect without account name or password!");
            }
            //create MimeMessage
            MimeMessage message = new MimeMessage(session);
            message.setFrom(new InternetAddress(prop.getProperty("sendFrom",prop.getProperty("senderAddress"))));
            message.setRecipients(Message.RecipientType.TO, prop.getProperty("recipientAddress"));
            message.setSubject(subjectInfo,"UTF-8");
            message.setContent(content, "text/html;charset=UTF-8");
            //new Date() means send email immediately
            message.setSentDate(new Date());
            transport.sendMessage(message,message.getAllRecipients());
            transport.close();
        } catch (Exception e) {
            LOGGER.error("send email failed.",e);
            LOGGER.error("email title:" + subjectInfo + ", email content:" + content);
        } finally{
            try {
                transport.close();
            } catch (MessagingException e) {
                e.printStackTrace();
            }
        }
    }




}

File: 数仓表结构/common/src/main/scala/com/weshare/utils/SendEmailUtil.scala
package com.weshare.utils

/**
 * created by chao.guo on 2020/8/29
 *
 **/
import java.util.{Date, Properties}

import javax.mail.internet.{InternetAddress, MimeMessage}
import javax.mail.{Message, Session, Transport}
import org.apache.hadoop.fs.{FileSystem, Path}
import org.apache.spark.sql.SparkSession

object SendEmailUtil {


  var session:Session =_;
  var pro:Properties= new  Properties();
  var transport:Transport=_;

  def initEmailSession(configPath:String,hdfs_master:String,isHa:Int) ={
    val conf = HdfsUtils.initConfiguration(hdfs_master,isHa)
    val inputSteam = FileSystem.get(conf).open(new Path(configPath))
    pro.load(inputSteam)
    inputSteam.close()
    session = Session.getInstance(pro)
    transport = session.getTransport
  }



  /** 发送邮件信息
   *
   * @param
   * @param subjectType
   */
  def  sendMessage(messag:String,subjectType:String): Unit ={
    val senderAccount = pro.getProperty("senderAccount")
    val senderPassword = pro.getProperty("senderPassword")
    if(!senderAccount.isEmpty && !senderPassword.isEmpty){
      transport.connect(senderAccount,senderPassword)
    }else{
      transport.connect();
    }
    // 拼装模版消息
    val msg = new MimeMessage(session)
    msg.setFrom(new InternetAddress(pro.getProperty("senderAddress")))

    msg.setFrom(new InternetAddress(pro.getProperty("sendFrom")))
    msg.addRecipients(Message.RecipientType.TO,pro.getProperty("recipientAddress"))
    msg.setSubject(subjectType, "UTF-8")
    //设置邮件正文
    msg.setContent(messag, "text/html;charset=UTF-8")
    //设置邮件的发送时间,默认立即发送
    msg.setSentDate(new Date())
    transport.sendMessage(msg,msg.getAllRecipients)
    transport.close() //关闭连接
}

  /**
   * 发送模版消息
   * @param messag
   * @param subjectType
   * @param recivers_email
   */
  def  sendMessage(messag:String,subjectType:String,recivers_email: String): Unit ={
    println("登陆邮件服务器")
    val senderAccount = pro.getProperty("senderAccount")
    val senderPassword = pro.getProperty("senderPassword")
    if(!senderAccount.isEmpty && !senderPassword.isEmpty){
      transport.connect(senderAccount,senderPassword)
    }else{
      transport.connect();
    }
    // 拼装模版消息
    val msg = new MimeMessage(session)
    msg.setFrom(new InternetAddress(pro.getProperty("senderAddress")))

    msg.setFrom(new InternetAddress(pro.getProperty("sendFrom")))
    if(org.apache.commons.lang3.StringUtils.isNoneBlank(recivers_email)){
      msg.addRecipients(Message.RecipientType.TO,recivers_email)
    }else{
      msg.addRecipients(Message.RecipientType.TO,pro.getProperty("recipientAddress"))
    }

    msg.setSubject(subjectType, "UTF-8")
    //设置邮件正文
    msg.setContent(messag, "text/html;charset=UTF-8")
    //设置邮件的发送时间,默认立即发送
    msg.setSentDate(new Date())
    transport.sendMessage(msg,msg.getAllRecipients)
    transport.close() //关闭连接
  }


}
